core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -3629,8 +3629,13 @@ private RecordCursorFactory generateSetFactory(
                     return null;
             }
         } catch (Throwable e) {
-            Misc.free(factoryA);
-            Misc.free(factoryB);
+            //factories could be already closed as part of 'union' factory in nested call
+            if (!factoryA.isClosed()) {
```

Comment:
why do we need to check if factory is closed when we made `close()`  idempotent?

Reply:
That's because close() is not idempotent when assertions are enabled . 
Alternative is to remove the assertion from close() .

Reply:
lets remove assertions!

Reply:
addressed in #bb302478